### DISCUSSION TITLE
What's the right way to programmatically set the selected RadioButton within a RadioSet?

### DISCUSSION THREAD

**Discussion by lauralindzey** [Textualize/textual#3983] [Q&A] [answered]
I have a RadioSet with a bunch of RadioButtons, e.g.:
~~~python
RadioSet(
    RadioButton("task 1a", classes="task1", value=True),
    RadioButton("task 1b", classes="task1", value=False),
    RadioButton("task 2a", classes="task2", value=False),
    RadioButton("task 2b", classes="task2", value=False),
    id = "target_radio_set"
)
~~~

Only a subset of them should be visible at any time, depending on the state of the rest of the program. I've figured out two ways to do that:
1) directly set `.display` on each RadioButton, directly update this when the program state changes.
~~~python
buttons = self.query("#target_radio_set>RadioButton.task1")
for obj in buttons:
        obj.display = True
~~~
2) assign RadioButtons to classes, and have another class that updates when the program state changes. Use tcss to manage `display` based on these two classes. e.g.:
```css
.task2_active #target_radio_set>RadioButton{
    display: none;
    &.task2 {
        display: block;
    }
}
```

However, once I have the proper radio buttons displaying, I want to either totally clear the RadioSet's state so that no buttons are active, or to default to the first displayed button being set to active. Is it possible to do either of these programmatically? (The important bit here is that if I query the RadioSet's state via `pressed_button`, it should never return a RadioButton that isn't currently displayed.)

I've tried setting `button.value=True` on an appropriate RadioButton itself, but that totally borks the RadioSet's functionality -- that button never turns off, even if I click another one in the set. I get similar behavior if I try `button.toggle()`. The only time setting `value` directly seems to work is in the actual constructor, when I'm creating the RadioSet. 

I also tried `button.emit(...)` based on [this discussion](https://github.com/Textualize/textual/discussions/186), but get the error "AttributeError: 'RadioButton' object has no attribute 'emit'"

**Comment by davep** [comment, marked as answer]
I feel you're trying to do a little more with `RadioSet` than it was intended to do. If I wanted a `RadioSet` whose content depended on the state of some other part of my application, I'd probably take an approach where I `remove` the existing one and `mount` fresh one in its place with the desired buttons.

As an illustration:

```python
from textual import on
from textual.app import App, ComposeResult
from textual.containers import Vertical
from textual.widgets import RadioSet, Select

class DynamicRadioSetApp(App[None]):

    def compose(self) -> ComposeResult:
        yield Select[int]((("The Beatles", 0), ("Radiohead", 1),("REM", 2)))
        with Vertical(id="band-picker"):
            yield RadioSet()

    @on(Select.Changed)
    async def pick_the_best_member(self, event: Select.Changed) -> None:
        set_container = self.query_one("#band-picker")
        await set_container.query_one(RadioSet).remove()
        members = []
        if event.value == 0:
            members = ["John", "Paul", "George", "Ringo"]
        elif event.value == 1:
            members = ["Thom", "Jonny", "Ed", "Philip"]
        elif event.value == 2:
            members = ["Bill", "Peter", "Mike", "Michael"]
        await set_container.mount(RadioSet(*members))

if __name__ == "__main__":
    DynamicRadioSetApp().run()
```

**Reply by davep** [reply]
Oh, and as for `emit`: you're looking at a pretty old issue there and it was removed [almost a year back](https://github.com/Textualize/textual/blob/main/CHANGELOG.md#removed-5).

**Reply by lauralindzey** [reply]
Thank you so much for the comprehensive answer @davep, that makes sense!
